src/dotty/tools/dotc/transform/InterceptedMethods.scala

```diff
@@ -96,14 +96,31 @@ class InterceptedMethods extends MiniPhaseTransform {
         s"that means the intercepted methods set doesn't match the code")
       tree
     }
-    lazy val Select(qual, _) = tree.fun
+    lazy val qual = tree.fun match {
+      case Select(qual, _) => qual
+      case ident @ Ident(_) =>
+        ident.tpe match {
+          case TermRef(prefix: TermRef, name) =>
+            tpd.ref(prefix)
+          case TermRef(prefix: ThisType, name) =>
+            tpd.This(prefix.cls)
+        }
+
+    }
     val Any_## = this.Any_##
     val Any_!= = defn.Any_!=
     val rewrite: Tree = tree.fun.symbol match {
       case Any_## =>
           poundPoundValue(qual)
       case Any_!= =>
-          qual.select(defn.Any_==).appliedToArgs(tree.args).select(defn.Boolean_!)
+        assert(tree.args.size == 1)
+        val lhs = qual.tpe
+        val rhs = tree.args.head.tpe
+        if (!(lhs <:< rhs || rhs <:< lhs) && rhs != defn.NullType && lhs != defn.NullType) {
```

Comment:
For example if I run `partest`without the `rhs != defn.NullType && lhs != defn.NullType` I get the following two warnings (with `partest-only-no-bootsrtap` I only get the second one):

```
[info] Test test.TestBCode.nullChecks started
-- Warning: 
class Foo {
  def foo(x: AnyRef): Int = {
    val bool = x == null
    if (x != null) 1
    else 0
  }
}
                  
5 |    if (x != null) 1
  |       ^^^^^^^^^^^
  |       comparing values of types Object and Null using `!=' will always yield true
```

```
[info] Test dotc.tests.repl_all started
input differs from transcript:
scala> val Const,x = 0
Const: Int = 0
x: Int = 0
scala> val (Const, List(`x`, _, a), b) = (0, List(0, 1337, 1), 2)
-- Warning: <console> --------------------------------------------------------------------------------------------------
6 |val (Const, List(`x`, _, a), b) = (0, List(0, 1337, 1), 2)
  |                 ^
  |         comparing values of types scala.collection.immutable.List[Int] and Null using `!=' will always yield true
a: Int = 1
b: Int = 2
scala> val a@b = 0
a: Int @unchecked = 0
b: Int @unchecked = 0
scala> :quit
```

Are those non nullable `AnyRef`s?